src/toxcore/tcp/server/server_ext.rs

```diff
@@ -1,41 +1,62 @@
 /*! Extension trait for run TCP server on `TcpStream` and ping sender
 */
 
-use std::io::{Error, ErrorKind};
+use std::io::{Error as IoError};
 use std::time::{Duration, Instant};
 
+use failure::Error;
 use futures::{future, Future, Sink, Stream};
 use futures::sync::mpsc;
 use tokio;
 use tokio::net::{TcpStream, TcpListener};
 use tokio::util::FutureExt;
 use tokio_codec::Framed;
 use tokio::timer::Interval;
+use tokio::timer::timeout::{Error as TimeoutError};
 
 use toxcore::crypto_core::*;
-use toxcore::io_tokio::IoFuture;
 use toxcore::tcp::codec::Codec;
 use toxcore::tcp::handshake::make_server_handshake;
 use toxcore::tcp::server::{Client, Server};
 
 /// Interval in seconds for Tcp Ping sender
 const TCP_PING_INTERVAL: u64 = 1;
 
+/// Error that can happen during TCP connection execution
+#[derive(Debug, Fail)]
+pub enum ConnectionError {
+    /// Error indicates that we couldn't get peer address
+    #[fail(display = "Failed to get peer address: {}", error)]
+    PeerAddrError {
+        /// Peer address error
+        error: IoError,
+    },
+    /// Error indicates that client stream was terminated. Should not happen
+    #[fail(display = "Client stream terminated")]
+    ClientStreamError,
+    /// Sending packet error. Can be either timeout or underlying sending error
+    #[fail(display = "Failed to send TCP packet: {}", error)]
+    SendPacketError {
+        error: TimeoutError<Error>
+    },
+}
+
 /// Extension trait for running TCP server on incoming `TcpStream` and ping sender
 pub trait ServerExt {
     /// Running TCP ping sender and incoming `TcpStream`. This function uses
     /// `tokio::spawn` inside so it should be executed via tokio to be able to
     /// get tokio default executor.
-    fn run(self: Self, listner: TcpListener, dht_sk: SecretKey) -> IoFuture<()>;
```

Comment:
Why did you replace `IoFuture` with `Box<...>`?

Reply:
Because now it's different type. Error is not `std::io::Error` but `failure::Error`. Later we could use existential types here.